src/parser/MetricCalculator.ts

```diff
@@ -29,30 +33,15 @@ export class MetricCalculator {
         ];
     }
 
-    calculateMetrics(parseFiles: ParseFile[]) {
-        const sourcesRoot = fs.realpathSync(this.config.sourcesPath);
-
-        console.log(
-            "\n\n",
-            "----- Calculating metrics for " + sourcesRoot + " recursively -----",
-            "\n\n"
-        );
-        console.log(" --- " + parseFiles.length + " files detected", "\n\n");
-
-        const fileMetrics = new Map<string, Map<string, MetricResult>>();
+    async calculateMetrics(parseFile: ParseFile) {
+        const metricResults = new Map<string, MetricResult>();
 
-        for (const parseFile of parseFiles) {
-            const metricResults = new Map<string, MetricResult>();
-            fileMetrics.set(parseFile.filePath, metricResults);
+        dlog(" ------------ Parsing File " + parseFile.filePath + ":  ------------ ");
 
-            console.log(" ------------ Parsing File " + parseFile.filePath + ":  ------------ ");
-
-            for (const metric of this.fileMetrics) {
-                const metricResult = metric.calculate(parseFile);
-                metricResults.set(metricResult.metricName, metricResult);
-            }
+        for (const metric of this.fileMetrics) {
+            const metricResult = await metric.calculate(parseFile);
```

Comment:
No, each promise is still awaited sequentially. The implementation should use something like `Promise.all`. My suggestion is just to limit the parallel work and that's why I suggest using p-map.

Reply:
I have tried to use p-map here, but that lead to the following runtime error:

> ...\metric-gardener\dist\parser\GenericParser.js:27
> const p_map_1 = __importDefault(require("p-map"));
>                                 ^
> 
> Error [ERR_REQUIRE_ESM]: require() of ES Module ...\metric-gardener\node_modules\p-map\index.js from ...\metric-gardener\dist\parser\GenericParser.js not supported.
> Instead change the require of index.js in ...\metric-gardener\dist\parser\GenericParser.js to a dynamic import() which is available in all CommonJS modules.
>     at Object.<anonymous> (...\metric-gardener\dist\parser\GenericParser.js:27:33) {
>   code: 'ERR_REQUIRE_ESM'
> }
> 
> Node.js v20.10.0

As far as I (not JavaScript-expert yet) understand, that means that the p-map dependency and the project use a different JavaScript module format?

Reply:
I have implemented the use of Promise.all(...) for now

Reply:
I have opened #35 for the p-map issue.

Reply:
Regarding the loop in MetricCalculator.ts, it was intentional to not make the calculation of the metrics of a single file asynchronous, as there is noting to do for each metric before the file is parsed first, as invoked by the first line of the calculate(...) function of the first metric, and only CPU-work afterwards, as the syntax tree is cached after being parsed once. It would also increase the number of promises, as for one of the promises for the single files in GenericParser.ts to complete when calling promises.all, these newly introduced promises for all single metrics for that single file would also have to be awaited. So I guessed that there is not much asynchrony to achieve here and that the overhead of handling and awaiting all these promises would not be worth it therefore, but I might be wrong with that, in which case I can change that. It would probably work well to use multithreading here to calculate the metrics in parallel once the file is parsed, but again, unsure if that would be worth the overhead.

Reply:
The `calculate()` method is frequently async (it is also possible to differentiate that to prevent the extra tick using `await` by checking if it is a promise or not on the return value sync), even if the file is cached. As such, it is a sequential execution that waits that tick. It is not a lot of time and in the current implementation, it will likely not make any difference. But I would definitely expect async tasks to be executed in parallel.

One design decision was that the calculate method accesses the files nodes from inside that method. What about injecting the nodes from outside? That would resolve the specific situation in a cleaner way.

Reply:
Ok. I unfortunately do not understand that in every detail ("frequently async", "return value sync"?), but I conclude that it can be worthwhile to make the calculation of metrics on a single file asynchronous. I therefore will implement that as soon as we clarify on how to await the multiple promises in GenericParser.ts (https://github.com/MaibornWolff/metric-gardener/pull/23#discussion_r1452625942), as currently, I would just implement awaiting the promises for every metric in the same way as I await the results for the files there.

Reply:
Yes, I have already thought of refactoring that the tree parsing is handled before calling the individual calculate methods. I will change that together with making that loop asynchronous.

Reply:
Should be done now